lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -2770,6 +2771,334 @@ def permutation_matrices(n):
+                [xlolims, ylolims, zlolims], [xuplims, yuplims, zuplims]):
+
+            dir_vector = art3d.get_dir_vector(zdir)
+            i_zdir = i_xyz[zdir]
+
+            if err is None:
+                continue
+
+            if not np.iterable(err):
+                err = [err] * len(data)
+
+            err = np.atleast_1d(err)
+
+            # arrays fine here, they are booleans and hence not units
+            lolims = np.broadcast_to(lolims, len(data)).astype(bool)
+            uplims = np.broadcast_to(uplims, len(data)).astype(bool)
+
+            nolims = ~(lolims | uplims)
+
+            # a nested list structure that expands to (xl,xh),(yl,yh),(zl,zh),
+            # where x/y/z and l/h correspond to dimensions and low/high
+            # positions of errorbars in a dimension we're looping over
+            coorderr = [
+                _unpack_errs(err * dir_vector[i], coord,
+                             ~lolims & everymask, ~uplims & everymask)
+                for i, coord in enumerate([x, y, z])]
+            (xl, xh), (yl, yh), (zl, zh) = coorderr
+
+            # draws capmarkers - flat caps othogonal to the error bars
+            if nolims.any() and capsize > 0:
+                lo_caps_xyz = _mask_lists(xl, yl, zl, nolims & everymask)
+                hi_caps_xyz = _mask_lists(xh, yh, zh, nolims & everymask)
+
+                # setting '_' for z-caps and '|' for x- and y-caps;
+                # these markers will rotate as the viewing angle changes
+                cap_lo = art3d.Line3D(*lo_caps_xyz, ls='',
+                                      marker=capmarker[i_zdir],
+                                      **eb_cap_style)
+                cap_hi = art3d.Line3D(*hi_caps_xyz, ls='',
+                                      marker=capmarker[i_zdir],
+                                      **eb_cap_style)
+                self.add_line(cap_lo)
+                self.add_line(cap_hi)
+                caplines.append(cap_lo)
+                caplines.append(cap_hi)
+
+            if (lolims | uplims).any():
+                limits = [
+                    _unpack_errs(err*dir_vector[i], coord, uplims, lolims)
+                    for i, coord in enumerate([x, y, z])]
+
+                (xlo, xup), (ylo, yup), (zlo, zup) = limits
+                lomask = lolims & everymask
+                upmask = uplims & everymask
+                lolims_xyz = _mask_lists(xlo, ylo, zlo, upmask)
+                uplims_xyz = _mask_lists(xup, yup, zup, lomask)
+                lo_xyz = _mask_lists(x, y, z, upmask)
+                up_xyz = _mask_lists(x, y, z, lomask)
+
+                x0, y0, z0 = x0, y0, z0 = np.concatenate([lo_xyz, up_xyz],
+                                                         axis=-1)
+                dx, dy, dz = np.concatenate([
+                    np.array(lolims_xyz) - np.array(lo_xyz),
+                    np.array(uplims_xyz) - np.array(up_xyz)], axis=-1)
```

Comment:
May as well halve the number of times you create the same arrays:
```suggestion
                lolims_xyz = np.array(_mask_lists(xlo, ylo, zlo, upmask))
                uplims_xyz = np.array(_mask_lists(xup, yup, zup, lomask))
                lo_xyz = np.array(_mask_lists(x, y, z, upmask))
                up_xyz = np.array(_mask_lists(x, y, z, lomask))

                x0, y0, z0 = np.concatenate([lo_xyz, up_xyz], axis=-1)
                dx, dy, dz = np.concatenate([lolims_xyz - lo_xyz, uplims_xyz - up_xyz], axis=-1)
```

Reply:
Thanks, that's a great suggestion. Fixed it in 5db6976.